cn/cn_dpnd/cn_dpnd_usr.c

```diff
@@ -0,0 +1,367 @@
+	 "Unload XDP program instead of loading"},
+
+	{{"quiet",       no_argument,		NULL, 'q' },
+	 "Quiet mode (no output)"},
+
+	{{"filename",    required_argument,	NULL,  1  },
+	 "Load program from <file>", "<file>"},
+
+	{{"progsec",    required_argument,	NULL,  2  },
+	 "Load program in <section> of the ELF file", "<section>"},
+
+	{{0, 0, NULL,  0 }}
+};
+
+int find_map_fd(struct bpf_object *bpf_obj, const char *mapname)
+{
+	struct bpf_map *map;
+	int map_fd = -1;
+
+	map = bpf_object__find_map_by_name(bpf_obj, mapname);
+        if (!map) {
+		fprintf(stderr, "ERR: cannot find map by name: %s\n", mapname);
+		goto out;
+	}
+
+	map_fd = bpf_map__fd(map);
+ out:
+	return map_fd;
+}
+
+static int __check_map_fd_info(int map_fd, struct bpf_map_info *info,
+			       struct bpf_map_info *exp)
+{
+	__u32 info_len = sizeof(*info);
+	int err;
+
+	if (map_fd < 0)
+		return EXIT_FAIL;
+
+        /* BPF-info via bpf-syscall */
+	err = bpf_obj_get_info_by_fd(map_fd, info, &info_len);
+	if (err) {
+		fprintf(stderr, "ERR: %s() can't get info - %s\n",
+			__func__,  strerror(errno));
+		return EXIT_FAIL_BPF;
+	}
+
+	if (exp->key_size && exp->key_size != info->key_size) {
+		fprintf(stderr, "ERR: %s() "
+			"Map key size(%d) mismatch expected size(%d)\n",
+			__func__, info->key_size, exp->key_size);
+		return EXIT_FAIL;
+	}
+	if (exp->value_size && exp->value_size != info->value_size) {
+		fprintf(stderr, "ERR: %s() "
+			"Map value size(%d) mismatch expected size(%d)\n",
+			__func__, info->value_size, exp->value_size);
+		return EXIT_FAIL;
+	}
+	if (exp->max_entries && exp->max_entries != info->max_entries) {
+		fprintf(stderr, "ERR: %s() "
+			"Map max_entries(%d) mismatch expected size(%d)\n",
+			__func__, info->max_entries, exp->max_entries);
+		return EXIT_FAIL;
+	}
+	if (exp->type && exp->type  != info->type) {
+		fprintf(stderr, "ERR: %s() "
+			"Map type(%d) mismatch expected type(%d)\n",
+			__func__, info->type, exp->type);
+		return EXIT_FAIL;
+	}
+
+	return 0;
+}
+#endif
+
+int cn_dp_assistant(int handle, bool ckmap)
+{
+    flow_ctx_t sendbuf;
+    int sockfd, rc;
+
+    if (handle < 0) {
+        printf("Invalid map handle, stop now\n");
+        return -1;
+    }
+
+    rc = 0;
+
+    struct sockaddr_in socket_receiver_address;
+    memset(&socket_receiver_address, 0, sizeof(socket_receiver_address));
+    socket_receiver_address.sin_family = AF_INET;
+    socket_receiver_address.sin_port = htons(PORT);
+    socket_receiver_address.sin_addr.s_addr = INADDR_ANY;
+    printf("DPA ready!\n");
+
+    /* Open RAW socket to send on */
+    if ((sockfd = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
+        printf("DPA failed to open raw socket! Err: %s\n", strerror(errno));
+        return 1;
+    }
+    printf("Successfully created a socket.\n");
+    for(;;){
+        //printf("One round in the forever for loop...\n");
+        dp_encap_opdata_t opdata;
+        unsigned char src_mac[6] = {108,221,238,0,0,44}; //"6c:dd:ee:00:00:2c"
```

Comment:
Should we keep these hard-coded lines?

Reply:
Ideally, they could be in a separate unit test code. Maybe we can do that later in another PR.

Reply:
I agree, let's do it then. 

lgtm.